src/GafferScene/MeshSegments.cpp

```diff
@@ -0,0 +1,277 @@
+	// that is indexed ) to uniform ( one value per face ).  We do this just by reading one vertex from
+	// each face.
+	vertexIdsIndex = 0;
+	for( int f : verticesPerFace )
+	{
+		uniformSegments.push_back( segments[ indices[ vertexIdsIndex ] ] );
+		vertexIdsIndex += f;
+	}
+}
+
+} // namespace
+
+size_t MeshSegments::g_firstPlugIndex = 0;
+
+GAFFER_NODE_DEFINE_TYPE( MeshSegments );
+
+MeshSegments::MeshSegments( const std::string &name )
+	:	ObjectProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringPlug( "segmentsPrimVar", Plug::In, "segments" ) );
+	addChild( new StringPlug( "indexPrimVar", Plug::In, "P" ) );
+}
+
+MeshSegments::~MeshSegments()
+{
+}
+
+Gaffer::StringPlug *MeshSegments::segmentsPrimVarPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *MeshSegments::segmentsPrimVarPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug *MeshSegments::indexPrimVarPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *MeshSegments::indexPrimVarPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+bool MeshSegments::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return
+		ObjectProcessor::affectsProcessedObject( input ) ||
+		input == segmentsPrimVarPlug() ||
+		input == indexPrimVarPlug()
+	;
+}
+
+void MeshSegments::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ObjectProcessor::hashProcessedObject( path, context, h );
+
+	segmentsPrimVarPlug()->hash( h );
+	indexPrimVarPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshSegments::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const std::string &segmentsPrimVar = segmentsPrimVarPlug()->getValue();
+	const std::string &indexPrimVar = indexPrimVarPlug()->getValue();
+	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject );
+	if( !mesh || !segmentsPrimVar.size() )
+	{
+		return inputObject;
+	}
+
+	auto it = mesh->variables.find( indexPrimVar );
+	if( indexPrimVar != "" && it == mesh->variables.end() )
+	{
+		throw IECore::Exception( "No primitive variable named \"" + indexPrimVar + "\"" );
+	}
+
+	const std::vector<int> &verticesPerFace = mesh->verticesPerFace()->readable();
+	IntVectorDataPtr uniformSegmentsData = new IntVectorData();
+	if( indexPrimVar == "" || it->second.interpolation == PrimitiveVariable::Interpolation::Vertex || it->second.interpolation == PrimitiveVariable::Interpolation::Varying )
+	{
+		if( it->second.indices )
+		{
+			throw IECore::Exception( "Vertex primitive variable " + indexPrimVar + " must NOT be indexed in order to use as indexPrimVar, because specifying a vertex primitive variable as indexPrimVar means \"Use the mesh's vertex indices\", and should not be affected by the primitive variable at all." );
```

Comment:
Maybe just "Vertex primitive variables with indices are not currently supported"?